### ISSUE TITLE
Lamda-lifting bug leading to crash.

### ISSUE THREAD

**Issue by yglukhov** [nim-lang/Nim#4551]
Not sure if its a good title for the bug, feel free to change it.
```nim
proc foo() =
    let arr = ["hi"]
    for i, v in arr:
        let bar = proc =
            echo v
        bar()
foo()
```
```
Traceback (most recent call last)
test.nim(7)              test
system.nim(2054)         foo
gc.nim(179)              asgnRefNoCycle
SIGBUS: Illegal storage access. (Attempt to read from nil?)
```

**Comment by RyanMarcus** [comment]
I hit this issue as well.

This executes **as expected**:

```
proc foo() =
    let arr = ["hi"]
    for i, v in arr:
        let localVar = v
        let bar = proc =
            echo localVar
        bar()
foo()
```

.. and this **fails**:

```
proc foo() =
    let arr = ["hi"]
    for i in arr:
        let bar = proc =
            echo i
        bar()
foo()
```

... so it looks like the error is with capturing a loop parameter directly within a lambda. The loop parameter doesn't appear to be captured correctly.

**Comment by RyanMarcus** [comment]
Confirmed fixed in `1a470d181854b078cf16e6cb6f7802fea8814680`

```
[ryan@rmarcus-zbook-arch nim]$ cat main.nim
proc foo() =
    let arr = ["hi"]
    for i in arr:
        let bar = proc =
            echo i
        bar()
foo()

[ryan@rmarcus-zbook-arch nim]$ ../Nim/bin/nim c -r main.nim
Hint: used config file '/home/ryan/Nim/config/nim.cfg' [Conf]
Hint: system [Processing]
Hint: main [Processing]
Hint:  [Link]
Hint: operation successful (10331 lines compiled; 0.104 sec total; 15.504MiB; Debug Build) [SuccessX]
Hint: /home/ryan/nim/main  [Exec]
hi
[ryan@rmarcus-zbook-arch nim]$ 
```
